llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-reverse.ll

```diff
@@ -229,10 +229,10 @@ define <32 x i8> @reverse_v32i8(<32 x i8> %a) {
 ; CHECK-LABEL: reverse_v32i8:
 ; CHECK:       # %bb.0:
 ; CHECK-NEXT:    li a0, 32
-; CHECK-NEXT:    lui a1, %hi(.LCPI12_0)
-; CHECK-NEXT:    addi a1, a1, %lo(.LCPI12_0)
 ; CHECK-NEXT:    vsetvli zero, a0, e8, m2, ta, ma
-; CHECK-NEXT:    vle8.v v12, (a1)
+; CHECK-NEXT:    vid.v v10
```

Comment:
I think I'm missing something, possibly two somethings.

1) It seems we already support fixed vector vector.reverse?  Do we current legalize to SHUFFLE_VECTOR or something?

2) Why is the prior lowering loading the constant from the constant pool?  The VID sequence here looks always profitable, even for generic shuffle lowering..

Reply:
For (2), I answered my own question.  Our constant lowering doesn't recognize reversed VID.

Reply:
For (1), it looks like SelectionDAGBuilder::visitVectorReverse unconditionally lowers to a SHUFFLE_VECTOR.

Reply:
> For (2), I answered my own question. Our constant lowering doesn't recognize reversed VID.

Is that true? There are many fixed vector reverse tests that already produce a vid.v+vrsub without going through VECTOR_REVERSE. They go through the BUILD_VECTOR lowering and produce vid.v+vrsub.

Reply:
Of the top of my head I think the vid sequence lowering has a check that the addend fits in uimm5

Reply:
Luke is right.  We do allow a negative step, it's just that the addend is "too large" in this case.  We should probably adjust our generic lowering as emitting the li seems like the right choice here.

Reply:
I played with relaxing the immediate constraint, and stumbled into a can of worms.  Basically, at high LMUL, our default VID strategy is generally not profitable.  For all but the most basic cases, we're better off narrowing the constant and then generating the VID sequence.  Doing that greatly reduces register pressure at high LMUL.  This is a existing problem with the code, but increasing the generality of the constant emission hits it at higher frequency.

Reply:
Mostly just to purge it out of my working memory, I wrote up some of the opportunities and missed optimizations I stumbled across when trying the VID narrowing.  I've uploaded them to my https://github.com/preames/public-notes/blob/master/llvm-riscv/vector-codegen.ll collection (starting w/ @vwaddu_vv).  As always, stuff I put here is low priority observations.